### PR TITLE
Fix primitive chunk free area error

### PR DESCRIPTION
**Pull request by singlecoder**
## Summary by CodeRabbit

- **New Features**
  - Introduced a dynamic text rendering feature in 3D environments, allowing text to appear one character at a time for a more engaging user experience.
  - Added a new configuration for enhanced handling of text-related cases within the end-to-end testing framework.

- **Bug Fixes**
  - Improved the merging logic for areas, ensuring that modifications to the areas occur under specific conditions, enhancing overall application stability and efficiency.

### CODE REVIEW

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts [hunk 1]
```diff
@@ -138,38 +138,49 @@ export class PrimitiveChunk {
...
+        pool.return(area);
+        const nextIndex = i + 1;
+        if (nextIndex < areaLen) {
+          const nextFreeArea = freeAreas[nextIndex];
+          if (end === nextFreeArea.start) {
+            // The cur free area after merge is to the left of the next free area and is connected.
+            curFreeArea.size += nextFreeArea.size;
+            freeAreas.splice(nextIndex, 1);
+            pool.return(nextFreeArea);
+          }
```
[the remark this answers is not in the document]
Comment (cptbtptpbcptdtptp) [reply packages/core/src/RenderPipeline/PrimitiveChunk.ts]: private _freeArea(insertArea: VertexArea): void {
    const areas = this.vertexFreeAreas;
    const areaLen = areas.length;
    if (areaLen === 0) {
      areas.push(insertArea);
      return;
    }

    const pool = PrimitiveChunk.areaPool;
    let insertStart = insertArea.start;
    let insertSize = insertArea.size;
    let insertEnd = insertStart + insertSize;
    for (let i = 0; i < areaLen; i++) {
      const area = areas[i];
      const areaStart = area.start;
      const areaEnd = areaStart + area.size;
      if (insertEnd < areaStart) {
        areas.splice(i, 0, insertArea);
        break;
      } else if (insertEnd === areaStart) {
        area.start = insertStart;
        area.size += insertSize;
        pool.return(insertArea);
        break;
      } else if (insertStart === areaEnd) {
        const nextArea = areas[i + 1];
        if (nextArea && nextArea.start === insertEnd) {
          area.size += insertSize + nextArea.size;
          pool.return(nextArea);
          areas.splice(i + 1, 1);
        } else {
          area.size += insertSize;
        }
        pool.return(insertArea);
        break;
      } else if (insertStart > areaEnd) {
        i + 1 === areaLen && areas.push(insertArea);
      }
    }
  }

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts [hunk 2]
```diff
@@ -138,38 +138,49 @@ export class PrimitiveChunk {
...
+        curFreeArea.start = start;
+        curFreeArea.size += size;
+        pool.return(area);
+        break;
+      } else if (start === curEnd) {
+        // The area to be freed is to the right of the current free area and is connected
+        curFreeArea.size += size;
+        pool.return(area);
+        const nextIndex = i + 1;
+        if (nextIndex < areaLen) {
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/PrimitiveChunk.ts:171]: size add twice?

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts [hunk 3]
```diff
@@ -138,38 +138,49 @@ export class PrimitiveChunk {
...
+            // The cur free area after merge is to the left of the next free area and is connected
+            curFreeArea.size += nextFreeArea.size;
+            freeAreas.splice(nextIndex, 1);
+            pool.return(nextFreeArea);
+          }
+        }
+        break;
+      } else if (start > curEnd) {
+        // The area to be freed is to the right of the current free area and is not connected
+        i + 1 === areaLen && freeAreas.push(area);
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/RenderPipeline/PrimitiveChunk.ts:183]:
```
  private _freeArea(insertArea: VertexArea): void {
    const areas = this.vertexFreeAreas;
    const areaLen = areas.length;
    if (areaLen === 0) {
      areas.push(insertArea);
      return;
    }

    const pool = PrimitiveChunk.areaPool;
    let insertStart = insertArea.start;
    let insertSize = insertArea.size;
    let insertEnd = insertStart + insertSize;
    for (let i = 0; i < areaLen; i++) {
      const area = areas[i];
      const areaStart = area.start;
      const areaEnd = areaStart + area.size;
      if (insertEnd < areaStart) {
        areas.splice(i, 0, insertArea);
        break;
      } else if (insertEnd === areaStart) {
        area.start = insertStart;
        area.size += insertSize;
        pool.return(insertArea);
        break;
      } else if (insertStart === areaEnd) {
        const nextArea = areas[i + 1];
        if (nextArea && nextArea.start === insertEnd) {
          area.size += insertSize + nextArea.size;
          pool.return(nextArea);
          areas.splice(i + 1, 1);
        } else {
          area.size += insertSize;
        }
        pool.return(insertArea);
        break;
      } else if (insertStart > areaEnd) {
        i + 1 === areaLen && areas.push(insertArea);
      }
    }
  }
```

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts (hunk 2 above)
Comment (singlecoder) [reply packages/core/src/RenderPipeline/PrimitiveChunk.ts:171]: yes

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts [hunk 4]
```diff
@@ -138,38 +138,49 @@ export class PrimitiveChunk {
...
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/PrimitiveChunk.ts:183]: Waste performance,use fallback mode: if not success, push it at end

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts [hunk 5]
```diff
@@ -138,39 +138,44 @@ export class PrimitiveChunk {
...
+        curFreeArea.size += size;
+        pool.return(area);
+        return;
+      } else if (start === curEnd) {
+        // The area to be freed is to the right of the current free area and is connected
+        curFreeArea.size += size;
+        pool.return(area);
+        const nextIndex = i + 1;
+        if (nextIndex < areaLen) {
+          const nextFreeArea = freeAreas[nextIndex];
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/RenderPipeline/PrimitiveChunk.ts:166]: nextFreeArea maybe not exit.

File: packages/core/src/RenderPipeline/PrimitiveChunk.ts (hunk 4 above)
Comment (singlecoder) [reply packages/core/src/RenderPipeline/PrimitiveChunk.ts]: done

**Pull request merged**